### ISSUE TITLE
GitLabList._data

### ISSUE THREAD

**Issue by fraser-brown-jdpa** [python-gitlab/python-gitlab#2948]
I've been using the GitlabList object to get security vulnerabilities reports from pipelines in some of my projects. I was using GitLab.http_list() returning a GitlabList object to iterate. However, the GitLab API doesn't return anything in the case that a Security report on a pipeline has expired. 

I've had a time trying to assure that GitlabList._data was not empty before beginning to iterate over it; I've had to do some janky work. I was hoping that a method could be added to access the protected attribute or check if it is empty.

Here is the [Vulnerability Findings](https://docs.gitlab.com/ee/api/vulnerability_findings.html#graphql---project-vulnerability-findings) API doc for reference.

**Comment by nejch** [comment]
Thanks for the report @fraser-brown-jdpa. This seems like you're using some of the lower level methods to paginate GraphQL responses - could you please provide a snippet of what you're trying to do?

`GitLabList` wasn't really meant to be used directly when it was created I assume, so might need some tweaking or documentation.

**Comment by fraser-brown-jdpa** [comment]
Yes, of course, here is a copy of the method I was hoping to do the "is empty" check in. Please don't judge me for the quality of this, it's still very much a work in progress. For clarity's sake, pipeline_url looks something like this "/api/v4/projects/{PROJECT_ID}/vulnerability_findings?pipeline_id={PIPELINE_ID}". The method is_gitlablist_empty() checks the protected attribute which I couldn't/can't do with mocking in Python, so I have had to patch that method to test this use case. 

```python
def _populate_report_by_project_and_pipeline(self, project: str, pipeline_id: int) -> ProjectReport:
    pipeline_url = self.__source_url.format(self.__git_url, self.__projects[project], pipeline_id)
    pipeline_report = self._git.http_list(path=pipeline_url, iterator=True, per_page=50)

    if self.is_gitlablist_empty(pipeline_report):
        raise ReportParsingException("No vulnerabilities found in report. Double check the report isn't expired.")

    return self.__parser.parse_report(pipeline_report, project)
```

**Comment by nejch** [comment]
@fraser-brown-jdpa ah ok, you're actually using the REST API then, not GraphQL, I was confused because of the link you sent.

I'm not sure exactly what your `is_gitlablist_empty` method does. Can you show a raw resposne (e.g. with curl) of what you get for expired ones? Maybe you just need to play around with the API parameters e.g. `scope`?

Usually, GitLab responds with the `X-Total-*` headers, so you could IMO just check for that, e.g. `if not pipeline_report.total: raise ...`. But this of course depends on whether GitLab reliably responds with the header, which we've seen for some endpoints it does not. Let us know though!

https://github.com/python-gitlab/python-gitlab/blob/8d74b889c55d18395c5e5006ba8f7db77892f603/gitlab/client.py#L1269

**Comment by fraser-brown-jdpa** [comment]
You're correct in the fact that .total is empty, however, ._data is not empty. Is this just going to be a limitation I have to deal with? This is how I have circumnavigated this issue, I needed this to test the scenario when the list is empty and I am unable to mock the protected attribute GitlabList._data:
```python
def is_gitlablist_empty(gitlablist: gitlab.GitlabList) -> bool:
        return len(gitlablist._data) < 1
```
When I debug into "your" method, GitlabList._query() I can see that X-Total in the response is absent, so GitlabList._total evaluates to None.

**Comment by nejch** [comment]
@fraser-brown-jdpa just to clarify, using private methods on a GitlabList is not really supported by us as it's internal implementation details. So you should really not rely on `_data` and `_query` but loop/iterate using `next()` and work with the data returned.

Are you saying that there's actual data returned by the server (i.e. vulnerability report json) but no `x-total` headers are present in the response?

**Comment by fraser-brown-jdpa** [comment]
Yes, that is what I'm saying, `x-total` is absent but there is content. I also want to clarify that I am not calling `GitlabList._query()`, I am using `._data` but that's why I made this issue, in hopes that a public method can reveal the protected attribute, or, since you seem to be aware of a few shortcomings with GitLab.


Sorry, if I haven't been very clear. Let me reword my problems. I have a use case where the content returned by the `Gitlab.http_list()` is empty. I created a check like you have pointed out using `.total()`, however, this does not work because `.total()` is always `None` for this API endpoint. This is due to the fact, like you pointed out, that `x-total` is absent in the response headers. I am petitioning you to create a method to access the protected attribute `._data` or implement something like this:
```python
def no_content(self) -> bool:
    if self._data is None:
        return True
    return len(self._data) < 1
```

Here is a snapshot from a debugger point in "your" `GitlabList._query()` method.
![image](https://github.com/user-attachments/assets/af36e013-ed33-425d-8dea-540399350a38)
